spark/src/main/scala/com/gu/tableversions/spark/SparkHiveMetastore.scala

```diff
@@ -0,0 +1,182 @@
+      tableLocation <- findTableLocation(table)
+      versionNumber = parseVersion(tableLocation)
+    } yield TableVersion(List(PartitionVersion(Partition.snapshotPartition, versionNumber)))
+
+    // Choose the calculation to perform based on whether we have a partitioned table or not
+    isPartitioned(table).flatMap(partitioned => if (partitioned) partitionedTableVersion else snapshotTableVersion)
+  }
+
+  override def update(table: TableName, changes: Metastore.TableChanges): F[Unit] =
+    changes.operations.map(appliedOp(table)).sequence.void
+
+  private def appliedOp(table: TableName)(operation: TableOperation): F[Unit] =
+    operation match {
+      case AddPartition(partitionVersion)                    => addPartition(table, partitionVersion)
+      case UpdatePartitionVersion(partitionVersion)          => updatePartitionVersion(table, partitionVersion)
+      case RemovePartition(partition)                        => removePartition(table, partition)
+      case UpdateTableLocation(tableLocation, versionNumber) => updateTableLocation(table, tableLocation, versionNumber)
+    }
+
+  private def addPartition(table: TableName, partitionVersion: PartitionVersion): F[Unit] = {
+
+    def addPartition(partitionExpr: String, partitionLocation: URI): F[Unit] = {
+      val addPartitionQuery =
+        s"ALTER TABLE ${table.fullyQualifiedName} ADD IF NOT EXISTS PARTITION $partitionExpr LOCATION '$partitionLocation'"
+      performUpdate(s"Adding partition to table ${table.fullyQualifiedName}", addPartitionQuery)
+    }
+
+    val partitionExpr = toHivePartitionExpr(partitionVersion.partition)
+
+    versionedPartitionLocation(table, partitionVersion).flatMap(location => addPartition(partitionExpr, location).void)
+  }
+
+  private def updatePartitionVersion(table: TableName, partitionVersion: PartitionVersion): F[Unit] = {
+
+    def updatePartition(partitionExpr: String, partitionLocation: URI): F[Unit] = {
+      val updatePartitionQuery =
+        s"ALTER TABLE ${table.fullyQualifiedName} PARTITION $partitionExpr SET LOCATION '$partitionLocation'"
+      performUpdate(s"Updating partition version of partition ${partitionVersion.partition}", updatePartitionQuery)
+    }
+
+    val partitionExpr = toHivePartitionExpr(partitionVersion.partition)
+
+    versionedPartitionLocation(table, partitionVersion).flatMap(location =>
+      updatePartition(partitionExpr, location).void)
+  }
+
+  private def removePartition(table: TableName, partition: Partition): F[Unit] = {
+    val partitionExpr = toHivePartitionExpr(partition)
+    val removePartitionQuery =
+      s"ALTER TABLE ${table.fullyQualifiedName} DROP IF EXISTS PARTITION $partitionExpr"
+    performUpdate(s"Removing partition $partition from table ${table.fullyQualifiedName}", removePartitionQuery)
+  }
+
+  private def versionedPartitionLocation(table: TableName, partitionVersion: PartitionVersion): F[URI] =
+    for {
+      tableLocation <- findTableLocation(table).map(new URI(_))
```

Comment:
I've not checked other usages, but might this always be a `URI`?

Reply:
I couldn't find a spec that explicitly says so, but from looking around I believe this is the case.

That said, it would be better if the `URI` values were produced at origin and not passed around as `String`s, so I'll fix that.